[B3_0] Review of wirecompression code

ibprovider opened the pull request:
Changed functions:
 - REMOTE_inflate
 - REMOTE_deflate
 - rem_port::initCompression
 - ServerAuth::authenticate
 - accept_connection
 - ConnectAuth::accept

dmitry-lipetsk committed:
Review of wirecompression code.

Changed functions:
 - REMOTE_inflate
 - REMOTE_deflate
 - rem_port::initCompression
 - ServerAuth::authenticate
 - accept_connection

dmitry-lipetsk committed:
Added comments

Function list:
 - ServerAuth::authenticate
 - accept_connection
 - ConnectAuth::accept

dmitry-lipetsk commented:
@AlexPeshkoff I understood that this PR is incorrect...

I was wrong about packet processing.

REMOTE_inflate may be called before setup of PORT_compressed flag.

So REMOTE_inflate looks on the port->port_compressed only and initCompression need call before authPort->send(send);

One question - why REMOTE_deflate checks both:
 (1) "port->port_compressed"
and
 (2) "port->port_flags & PORT_compressed" ?

Checking "port->port_flags & PORT_compressed" is not enough?

dmitry-lipetsk committed:
Restore of work

- REMOTE_inflate
   - restored usage of port->port_compressed as compression flag

- REMOTE_deflate
   - deleted wrong assert

- ServerAuth::authenticate
- accept_connection
- ConnectAuth::accept
   - restored old behaviour and added comments

dmitry-lipetsk committed:
formatting [tabs]

AlexPeshkoff commented:
I do not see what problems does suggested PR solve.

aafemt commented:
> I do not see what problems does suggested PR solve.


If you have a circular reference of such pointers, then deleting of pointed object may 
cause this object to be deleted as well, crashing at member access.

This is crazy and should never happen, but CORE-6224 imply that it is possible.

AlexPeshkoff commented:
@aafemt looks like you post at the wrong place

aafemt commented:
Yep, sorry.

dmitry-lipetsk committed:
Merge remote-tracking branch 'firebird/B3_0_Release' into B3_0-D2020_01_09-wirecompression_review_01

dmitry-lipetsk commented:
This PR does not solve the problems. And, I hope, does not create them.

This PR enhances code cleanliness and quality.

AlexPeshkoff commented on src/remote/remote.cpp:
```diff
@@ -1425,7 +1425,13 @@ bool REMOTE_inflate(rem_port* port, PacketReceive* packet_receive, UCHAR* buffer
+
+	fb_assert(port->port_compressed);
+
+	// Attention: PORT_compression flag may still not setup!
```
All this block does not look good.
1. Adding { } for single line block does not follow FB code standard (both styles possible but no cleanup from one to other).
2. "assert(expr)" right after "if (!expr) return" look wierd
3. Syntax error in comment, not PORT_compressION but PORT_compressED

AlexPeshkoff commented on src/remote/remote.cpp:
```diff
@@ -1503,19 +1509,17 @@ bool REMOTE_deflate(XDR* xdrs, ProtoWrite* proto_write, PacketSend* packet_send,
 		return proto_write(xdrs);
+	}//if
+
+	fb_assert(port->port_compressed);
```
Same issue (do not mention them later) with { } around single line comment.

AlexPeshkoff commented on src/remote/server/server.cpp:
```diff
@@ -709,11 +709,30 @@ class ServerAuth : public GlobalStorage, public ServerAuthBase
+				{
+					// Configuration of port for compression usage (1-3).
+					fb_assert(send->p_acpt.p_acpt_type & pflag_compress);
+
```
Once again assert() right after 'if'

AlexPeshkoff commented on src/remote/server/server.cpp:
```diff
@@ -709,11 +709,30 @@ class ServerAuth : public GlobalStorage, public ServerAuthBase
+					// 3. Enabled deflate.
 					authPort->port_flags |= PORT_compressed;
+				}//else
+
```
You remove duplication of condition check but add one for function call. Very doubtfully...
And code becomes 5 times longer than before. That's also inconvenient - we still do not have video walls from Bradbury's 451 as monitors :) Personally I dislike this change.

AlexPeshkoff commented:
In that case please close it and create PR for master, cleaning branches in this way does not make much sense.

dmitry-lipetsk commented:
ok
